src/Entity/AlertBannerEntity.php

```diff
@@ -406,7 +409,21 @@ public function getCacheContexts() {
       $this->addCacheContexts($contexts);
     }
 
-    return parent::getCacheContexts();
+    return Cache::mergeContexts(parent::getCacheContexts(), $contexts);
```

Comment:
I'm not entirely sure of the proper way to do this, but what we're assuming here is that all the cache contexts for an alert banner are entirely defined in this method or on the parent. If we wanted to add cache tags to an alert banner somewhere else they could be added to the `cacheContexts` property using the `addCacheContexts` method, but this won't work any longer.

I think it's safer to add the new `$contexts` using to the entity using the `addCacheContexts` method and then either return the `cacheContexts` property or `parent::getCacheContexts()` (which will do the same).

It _may_ be better to actually move the logic here out of the `getCacheContexts` method and have it run by the constructor to ensure it's only run once. There's the possibility that it will be called multiple times, which won't do any harm, but is inefficient.

Reply:
> If we wanted to add cache tags to an alert banner somewhere else they could be added to the cacheContexts property using the addCacheContexts method, but this won't work any longer.

As far as I understand, [parent::getCacheContexts()](https://git.drupalcode.org/project/drupal/-/blob/9.2.x/core/lib/Drupal/Core/Cache/CacheableDependencyTrait.php#L62) returns ```$cacheContexts```.  This means AlertBannerEntity::getCacheContexts() should always return whatever has been added through ```addCacheContexts()```

Reply:
I actually preferred ```Cache::mergeContexts(parent::getCacheContexts(), $contexts);```.  That seem to be what most modules do.  [Example from the book module](https://git.drupalcode.org/project/drupal/-/blob/9.2.x/core/modules/book/src/Plugin/Block/BookNavigationBlock.php#L186).  Not a big deal though.

Reply:
I've pushed changes with moving them to the consturcter...
However is as @Adnan-cds says, if the standard Drupal way is to use the `Cache::mergeContexts(parent::getCacheContexts(), $contexts);` from `getCacheContexts()` then I will revert it to that.

Reply:
Looks good to me.  Thanks.

> It may be better to actually move the logic here out of the getCacheContexts method and have it run by the constructor to ensure it's only run once.

I am not ignoring Steve's concern.  If it later turns out that this is indeed slowing down things then we can also consider using the good old [drupal_static()](https://api.drupal.org/api/drupal/core%21includes%21bootstrap.inc/function/drupal_static/9.1.x).

Reply:
I think this is fine. I don't fully understand why core doesn't make more use of the `addCacheContexts` property, but happy to go with this.